.github/workflows/haskell-ci.yml

```diff
@@ -141,6 +141,16 @@ jobs:
           xz -d < cabal-plan.xz > $HOME/.cabal/bin/cabal-plan
           rm -f cabal-plan.xz
           chmod a+x $HOME/.cabal/bin/cabal-plan
+          cabal-plan --version
+      - name: install cabal-docspec
+        run: |
+          mkdir -p $HOME/.cabal/bin
+          curl -sL https://oleg.fi/cabal-docspec-0.0.0.20201228.xz > cabal-docspec.xz
+          echo 'edf15d3a0b46fc57fb1bdb0ae5e87abdc5c3b7dfff8655a958ef038d4c2f17ab  cabal-docspec.xz' | sha256sum -c -
+          xz -d < cabal-docspec.xz > $HOME/.cabal/bin/cabal-docspec
+          rm -f cabal-docspec.xz
+          chmod a+x $HOME/.cabal/bin/cabal-docspec
+          cabal-docspec --version
```

Comment:
Admittedly, I was under the impression that `cabal-docspec` was something that you ran as a dependency in the test suite, not as an external executable. What advantages does this provide over simply running the `doctest` executable? (I apologize if this is a silly question—I still don't fully understand the internal workings of `cabal-docspec`.)

Reply:
`doctest` executable have to be compiled with GHC you want to test with. (So you cannot *simply* download it, problem that `ghc-mod`, `hie` etc have, AFAIK). `doctest` uses `ghc` library to both extract the docstrings as well as running own `ghci` (it doesn't call to external processes).

On the other hand, `cabal-docspec` just works with any GHC. A bit like `ghcid` doesn't really care about GHC version you use.

Another difference: because `cabal-docspec` isn't limited by "have to be fast to compile" (because we can download it), it can be fancier. Important fancy feature is that it can self configure. In short, the (kind of) logic we had in `cabal-doctest` custom build script is now moved into `cabal-docspec` executable. (We can learn enough from `plan.json` `cabal-install` generates).

Reply:
One concrete example of problems with `.ghc.environment` files and `doctest` is visibility of packages.

`haskell-ci` has a perl-hack to allow filtering packages, e.g. to remove `base-compat` from the environment files (which contains transitive dependencies) for packages which use `base-compat-batteries`.

`cabal-docspec` can be more precise (haskell vs. bash+perl :) )

Reply:
> Another difference: because `cabal-docspec` isn't limited by "have to be fast to compile" (because we can download it), it can be fancier.

To clarify, are you envisioning making binary distributions of `cabal-docspec` available in the future so that one can avoid needing to build it from source?

> Important fancy feature is that it can self configure. In short, the (kind of) logic we had in `cabal-doctest` custom build script is now moved into `cabal-docspec` executable. (We can learn enough from `plan.json` `cabal-install` generates).

Interesting. When does `plan.json` get generated in this script?

> One concrete example of problems with `.ghc.environment` files and `doctest` is visibility of packages.
> 
> `haskell-ci` has a perl-hack to allow filtering packages, e.g. to remove `base-compat` from the environment files (which contains transitive dependencies) for packages which use `base-compat-batteries`.
> 
> `cabal-docspec` can be more precise (haskell vs. bash+perl :) )

Ah, I had no idea that this hack was necessary. Indeed, that's a very good reason to use something more sophisticated than a bare `doctest` invocation.

Reply:
> To clarify, are you envisioning making binary distributions of cabal-docspec available in the future so that one can avoid needing to build it from source?

Yes.

> Interesting. When does plan.json get generated in this script?

It's a side-effect of doing about anything with `cabal-install`.

Even `cabal build --dry` is enough, though then some units (e.g. local) will be missing.

Reply:
Huh, today I learned that `cabal-install` [automatically generates a `plan.json`](https://cabal.readthedocs.io/en/3.4/nix-local-build.html?highlight=plan.json#caching) file that is located in `dist-newstyle`. Neat!

Reply:
Yes, it's what makes `cabal-plan` able to do all the things it does.